[wasm][http] Add support for Blob URLs

kjpou1 opened the pull request:
`blob:` should be valid URL schemes.

A valid blob scheme can be represented as follows:  
   - new Uri("blob:http://host/absolute/")
   - new Uri("blob:https://mdn.mozillademos.org/ca45b575-6348-4d3e-908a-3dbf3d146ea7")

---

Multiple tests were added

Before modification the following error was thrown:  ` System.ArgumentException : Only 'http' and 'https' schemes are allowed. (Parameter 'requestUri')`  

```
test[0]
        [FAIL] System.Runtime.InteropServices.JavaScript.Http.Tests.HttpRequestMessageTest.Ctor_AbsoluteBlobStringUri_CorrectValues 
  dbug: test[0]
        System.ArgumentException : Only 'http' and 'https' schemes are allowed. (Parameter 'requestUri')      
  dbug: test[0]
           at System.Net.Http.HttpRequestMessage.InitializeValues(HttpMethod method, Uri requestUri)
  dbug: test[0]
           at System.Net.Http.HttpRequestMessage..ctor(HttpMethod method, String requestUri)
  dbug: test[0]
           at System.Runtime.InteropServices.JavaScript.Http.Tests.HttpRequestMessageTest.Ctor_AbsoluteBlobStringUri_CorrectValues()
  dbug: test[0]
           at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)   
  dbug: test[0]
        Finished System.Private.Runtime.InteropServices.JavaScript.Tests, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51

```
---

Modify exception message to throw `Only 'http' 'https' and 'blob' schemes are allowed.` for browser.

---

Should be back ported to 5.0

close: https://github.com/dotnet/runtime/issues/41603

kjpou1 committed:
[wasm][http] Add support for Blob URLs

kjpou1 committed:
Modify the string to read `Only 'http' 'https' and 'blob' schemes are allowed.` for browser

kjpou1 committed:
Add more tests

kjpou1 committed:
Add blob Uri marshal test

lewing reviewed (commented):
I'm confused by the tests, is createObjectURL not working?

kjpou1 commented:
createObjectURL does not exist in the javascript engine.  Thus the hard coding.  Left it in because I already had it coded.

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/Http/HttpRequestMessageTest.cs:
```diff
@@ -174,7 +252,47 @@ public void Properties_SetOptionsAndGetTheirValue_Set_FetchOptions()
+        }
+
+        [Fact]
+        public void Properties_BlobSetOptionsAndGetTheirValue_Set_FetchOptions()
```
I left them separate to make sure it was tests specific to `Blob`.

kjpou1 replied:
Addressed

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/Http/HttpRequestMessageTest.cs:
```diff
@@ -346,6 +547,93 @@ public void ToString_DefaultAndNonDefaultInstance_DumpAllFields()
+        public void ToString_Blob_DefaultAndNonDefaultInstance_DumpAllFields()
+        {
+            var rm = new HttpRequestMessage();
+            string expected =
```
I guess we should also change the tests here as well right?  https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/tests/FunctionalTests/HttpRequestMessageTest.cs

radical replied:
Ah ok, then I guess that's the way it is being tested. Up to you then, what style to follow for this test.

kjpou1 replied:
Addressed

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/Http/HttpRequestMessageTest.cs:
```diff
@@ -346,6 +547,93 @@ public void ToString_DefaultAndNonDefaultInstance_DumpAllFields()
+            Runtime.InvokeJS(@"
+                function typedArrayToURL(typedArray, mimeType) {
+                    //return URL.createObjectURL(new Blob([typedArray.buffer], {type: mimeType}))
+                    return 'blob:https://mdn.mozillademos.org/ca45b575-6348-4d3e-908a-3dbf3d146ea7';
```
When creating a url like this `URL.createObjectURL` it will always be different so the url will never be fixed.  Just so happens to be fixed here because I took the test from  https://developer.mozilla.org/en-US/docs/Web/API/Blob and the URL.createObjectURL does not exist in the JS engine.

radical replied:
I completely missed the comment right above that line. Maybe add a comment to explain your rationale, for people like me!

kjpou1 replied:
Comments added

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/Http/HttpRequestMessageTest.cs:
```diff
@@ -346,6 +547,93 @@ public void ToString_DefaultAndNonDefaultInstance_DumpAllFields()
+                    //return URL.createObjectURL(new Blob([typedArray.buffer], {type: mimeType}))
+                    return 'blob:https://mdn.mozillademos.org/ca45b575-6348-4d3e-908a-3dbf3d146ea7';
+                }
+                const bytes = new Uint8Array(59);
+                for(let i = 0; i < 59; i++) {
+                    bytes[i] = 32 + i;
+                }
+                const url = typedArrayToURL(bytes, 'text/plain');
```
None of this is actually needed at all and it could just pass `blob:https://mdn.mozillademos.org/ca45b575-6348-4d3e-908a-3dbf3d146ea7` directly.

kjpou1 replied:
Addressed

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-http-blob-urls

kjpou1 committed:
Address review comments to add comments within code about intentions.

kjpou1 committed:
Multiple review comments.

- Split Fact into theory instead of specific test for blob.
- Split large test into multiple tests as well as mark it Theory where appropriate.

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/Http/HttpRequestMessageTest.cs:
```diff
@@ -346,6 +547,93 @@ public void ToString_DefaultAndNonDefaultInstance_DumpAllFields()
         }
 
+        [Fact]
+        public void ToString_Blob_DefaultAndNonDefaultInstance_DumpAllFields()
```
Addressed

radical commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs:
```diff
@@ -189,7 +189,11 @@ private void InitializeValues(HttpMethod method, Uri? requestUri)
             }
             if ((requestUri != null) && (requestUri.IsAbsoluteUri) && (!HttpUtilities.IsHttpUri(requestUri)))
```
`HttpUtilities.IsHttpUri(requestUri)` - we return true now if `scheme==blob`. What happens if we have a url like `blob:not-http://foo/bar`? Should `false` be returned in that case?

kjpou1 replied:
No as this is created from the javascript environment normally with URL.createURLObject.  The browser environment in this case will handle the error if something went wrong.

radical commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs:
```diff
@@ -189,7 +189,11 @@ private void InitializeValues(HttpMethod method, Uri? requestUri)
             }
             if ((requestUri != null) && (requestUri.IsAbsoluteUri) && (!HttpUtilities.IsHttpUri(requestUri)))
             {
+#if TARGETS_BROWSER
+                throw new ArgumentException(SR.net_http_client_http_browser_baseaddress_required, nameof(requestUri));
+#else
                 throw new ArgumentException(SR.net_http_client_http_baseaddress_required, nameof(requestUri));
+#endif
```
This same change is required in the `RequestUri` getter too, IIUC. And tests needed for that.

radical replied:
And is this same change required in [HttpClient:CheckBaseAddress](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L946-L949) too? Tests needed for that too.

kjpou1 replied:
Messages changed.

kjpou1 replied:
HttpClient can not be tested as it will not run in a non browser environment right now.  This will be tested when we get the https://github.com/dotnet/runtime/tree/master/src/libraries/System.Net.Http/tests running in the browser.

radical commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpUtilities.cs:
```diff
@@ -37,6 +41,11 @@ internal static bool IsHttpUri(Uri uri)
             string.Equals(scheme, "wss", StringComparison.OrdinalIgnoreCase);
 
+#if TARGETS_BROWSER
+        internal static bool IsNonSecureBlobScheme(string scheme) =>
```
nit: `IsBlobScheme`, assuming there is no secure blob scheme.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-http-blob-urls

kjpou1 committed:
Rename

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/Http/HttpRequestMessageTest.cs:
```diff
@@ -308,35 +327,53 @@ public void Method_SetToNull_ThrowsArgumentNullException()
 
         [Fact]
-        public void ToString_DefaultAndNonDefaultInstance_DumpAllFields()
+        public void ToString_DefaultPutInstance_DumpAllFields()
```
maybe just default

kjpou1 committed:
Add message to other validations as well

kjpou1 committed:
Change name

radical reviewed (approved):
LGTM! Though keep in mind that I'm new to this part of the code base!

kjpou1 commented:
Thanks for the review @radical .  I would still like a review from the network people as well.

stephentoub commented on src/libraries/System.Net.Http/src/Resources/Strings.resx:
```diff
@@ -204,6 +204,9 @@
     <value>Only 'http' and 'https' schemes are allowed.</value>
   </data>
+  <data name="net_http_client_http_browser_baseaddress_required" xml:space="preserve">
+    <value>Only 'http' 'https' and 'blob' schemes are allowed.</value>
```
```suggestion
    <value>Only 'http', 'https', and 'blob' schemes are allowed.</value>
```

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs:
```diff
@@ -945,7 +945,11 @@ private static void CheckBaseAddress(Uri? baseAddress, string parameterName)
 
             if (!HttpUtilities.IsHttpUri(baseAddress))
             {
+#if TARGETS_BROWSER
```
I'd like to avoid the proliferation of such ifdefs.  We've generally tried to do this with partial files, e.g. this would call a method `ThrowInvalidUril` which would be defined appropriately in the relevant build.

Alternatively, I'd be fine if we just changed the shared error message to be `The 'blah' scheme is not supported.` rather than having two different messages that list what is supported.

kjpou1 replied:
I like the more detailed one so left it in but moved the message definition to partial file as described.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs:
```diff
@@ -103,7 +103,11 @@ public HttpMethod Method
             {
                 if ((value != null) && (value.IsAbsoluteUri) && (!HttpUtilities.IsHttpUri(value)))
                 {
+#if TARGETS_BROWSER
```
Ditto

kjpou1 replied:
I like the more detailed one so left it in but moved the message definition to partial file as described.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs:
```diff
@@ -189,7 +193,11 @@ private void InitializeValues(HttpMethod method, Uri? requestUri)
             }
             if ((requestUri != null) && (requestUri.IsAbsoluteUri) && (!HttpUtilities.IsHttpUri(requestUri)))
             {
+#if TARGETS_BROWSER
```
Ditto

kjpou1 replied:
I like the more detailed one so left it in but moved the message definition to partial file as described.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpUtilities.cs:
```diff
@@ -26,7 +26,11 @@ internal static bool IsHttpUri(Uri uri)
             IsSupportedSecureScheme(scheme);
 
         internal static bool IsSupportedNonSecureScheme(string scheme) =>
```
Another case where I'd prefer if we moved this to two partial files, with each doing the desired checks, rather than proliferating use of TARGETS_BROWSER everywhere.  We've avoided that for platforms in general.

steveisok replied:
@kjpou1 In following what we've done before, you'd create `HttpUtilities.Browser.cs` and `HttpUtilities.AnyOS.cs`.  You'll also need to update the csproj accordingly.  There are numerous examples to draw from.

kjpou1 replied:
Moved this to separate partial files.

kjpou1 committed:
Update src/libraries/System.Net.Http/src/Resources/Strings.resx

Co-authored-by: Stephen Toub <stoub@microsoft.com>

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-http-blob-urls

kjpou1 committed:
Create partial classes for HttpUtilities to replace the proliferating use of TARGETS_BROWSER in the sources.

kjpou1 committed:
Define and call `HttpUtilities.InvalidUriMessage` to provide the invalid message to be thrown.

kjpou1 committed:
Fix CI Build

kjpou1 committed:
Rename modules to follow current standards

kjpou1 commented:
Can we get another round of eyes here?  @stephentoub @karelz

stephentoub reviewed (approved):
Thanks

lewing commented:
/backport to release/5.0-rc2

kjpou1 commented:
/backport to release/5.0-rc2
